services/src/main/java/org/keycloak/forms/login/freemarker/LoginFormsUtil.java

```diff
@@ -63,6 +63,7 @@ public static List<IdentityProviderModel> filterIdentityProviders(Stream<Identit
             return providers
                     .filter(p -> { // Filter current IDP during first-broker-login flow. Re-authentication with the "linked" broker should not be possible
                         if (existingIdp == null) return true;
+                        if (context.getUser() == null) return true;
```

Comment:
If you look carefully at the next filter there is a check for:
```
if (federatedIdentities == null) return true;
``` 
federatedIdentities will only be `null` if the `context.getUser()` is null as you can see from the check in line 55. Your change here has no effect.

Reply:
Also, based on the comment: `Re-authentication with the "linked" broker should not be possible`

I think this was the intended behavior

Reply:
But those providers will be filtered _out_ in this step, they won't even reach the lines below.

'cause the:
 - `Objects.equals(p.getAlias(), existingIdp.getAlias())` will be true (the alias of the cached-but-not-used provider will be one of the aliases of the available providers)
 -  and we negate this expression here
 - so the cached-but-not-used provider will be filtered out by this filter and never reach the next one

My fix is an early-exit, so if there's no user logged in then we can reach the filter below.

Reply:
> I think this was the intended behavior

Please, check the video I've attached to the issue :slightly_smiling_face: 

I do not finish the authentication flow, so there's no user authenticated. No user linked with the Identity Provider at all.

Reply:
I saw the video. Are you sure when you get the review profile step the user is not yet created / authenticated ? In this case, how does keycloak knows which user is modifying their profile

Reply:
> No user created, so there's no-one linked with the Identity Provider with the given ID after the steps.

Ok that answers the question. I think it makes sens but maybe the solution could be just modifying the `First Broker` Flow by either:
- making the `Review profile` step after the user creating/linking => which is the natural order of things
- making the review profile `Update profile on first login` value to `missing` instead of `on`.

Reply:
>  Are you sure when you get the review profile step the user is not yet created / authenticated ? In this case, how does keycloak knows which user is modifying their profile

Yeah, I'm pretty sure about that. You can test on a keycloak instance, that no user is created if you press the back button on the review page.

No user can be created, 'cause there's a step for that it the First Broker Login flow:

![image](https://user-images.githubusercontent.com/6528869/189731845-cd85101a-ca8f-4133-941d-d88a1fd3f586.png)

The user is created in the _create user if unique_ authenticator. (See the code here: [IdpCreateUserIfUniqueAuthenticator::75](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/authentication/authenticators/broker/IdpCreateUserIfUniqueAuthenticator.java#L75)) And no user is created when there's a duplication AND the duplicated users are forbidden in the realm settings.

> - making the Review profile step after the user creating/linking => which is the natural order of things
> - making the review profile Update profile on first login value to missing instead of on.

I think you cannot do that. Consider the following example: given User John Doe with the following accounts:
 - Facebook, where he uses his childhood's email address: jhonny9310@aol.net
 - Google, where he has a more formal e-mail address: john.doe@gmail.com
 - And he has an Apple account, where he has the private relay setting on, so the e-mail address Apple gives to keycloak is a randomly generated string 

He registered with his Google account months ago, and when he next visits our site he does not remember which IDP he used. He clicks on the Apple login (he opened the site from an iPhone, so Apple login is a convenient one). Keycloak presents him the following info:
 - this IDP is not yet linked with an account
 - and he can review the e-mail address he'd like to use

So John has the ability to change the e-mail address to his main one: john.doe@gmail.com And from there Keycloak takes care of merging his Apple ID to the pre-existing account. I'm pretty sure this is the expected behavior from the Keycloak team (see the flow, there are multiple verification options, and stuff).

 - If I'd change the First broker login flow in my installation (eg: move the review step after the user creation), then I'd be divert from upstream AND lose this merging functionality
 - If I'd make this change in the upstream keycloak, then the upstream keycloak lose this merging functionality
   - Or if we'd take care of the merging, we'd introduce a lot of fake user creation. There'd be created a user for the apple account in the previous example, AND we should merge the two accounts when John changes the email address in one of the accounts... But what should we do if he changes the e-mail address minutes/months after the registration? What should our clients do? Should _all_ Keycloak clients prepare for merged users?

Imagine a subscription website which uses Keycloak as the authentication. Should this website merge the users' subscriptions after the users had been merged? Who takes care of the refunds, and so on... Huge mess.

I'm pretty sure that we (the Keycloak team and myself) do not want to change how the First Broker login flow works.

And I'm pretty sure it's a bug, even if it's a small one, that the IDP button disappears sometimes.

I've tested it: this change fixes this bug, the only thing I'm not sure of, whether there are security issues with it or not. I'm not _THAT_ familiar with OAuth & Keycloak yet.